homeassistant/components/sonarr/calendar.py

```diff
@@ -0,0 +1,78 @@
+"""Support for Sonarr calendar."""
+from __future__ import annotations
+
+from datetime import datetime, timedelta
+
+from homeassistant.components.calendar import CalendarEntity, CalendarEvent
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DOMAIN
+from .coordinator import CalendarDataUpdateCoordinator
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up Sonarr calendar based on a config entry."""
+    coordinator = hass.data[DOMAIN][entry.entry_id]["upcoming"]
+    async_add_entities([SonarrCalendarEntity(coordinator, "Sonarr Episodes")])
+
+
+async def get_sonarr_episode_events(
+    coordinator: CalendarDataUpdateCoordinator,
+) -> list[CalendarEvent]:
+    """Update the list of upcoming episodes."""
+    calendar_entries: list[CalendarEvent] = []
+    for episode in coordinator.data:
+        episode_endtime_utc = episode.airDateUtc + timedelta(
+            minutes=episode.series.runtime
+        )
+        calendar_entries.append(
+            CalendarEvent(
+                summary=episode.series.title,
+                description=episode.title,
+                location=episode.series.network,
+                start=episode.airDateUtc,
+                end=episode_endtime_utc,
+            )
+        )
+    return calendar_entries
+
+
+class SonarrCalendarEntity(CalendarEntity):
+    """Representation of a Sonarr Calendar element."""
+
+    coordinator: CalendarDataUpdateCoordinator
+    _events: list[CalendarEvent]
+
+    def __init__(self, coordinator: CalendarDataUpdateCoordinator, name: str) -> None:
+        """Initialize Sonarr calendar."""
+        self.coordinator = coordinator
+        self._attr_name = name
+        self._events = []
+
+    @property
+    def event(self) -> CalendarEvent | None:
+        """Return the next upcoming event."""
+        if len(self._events) > 0:
+            return self._events[0]
+        return None
+
+    async def async_get_events(
+        self,
+        hass: HomeAssistant,
+        start_date: datetime,
+        end_date: datetime,
+    ) -> list[CalendarEvent]:
+        """Return calendar events within a datetime range."""
+        self._events = await get_sonarr_episode_events(self.coordinator)
```

Comment:
I'm missing the datetime range filter of the events here.

Reply:
The docstring is inaccurate, the events come from the Coordinator which has a configured number of days via config flow. 

Do you suggest just updating the docstring, or making an new coordinator?

Reply:
This method will be called by the calendar websocket api with start and end datetime parameters to filter the events. I would expect us to honor those parameters here.

Reply:
Just a correction of what I said earlier: This api is not a called from a websocket command but an http view.